webapp/channels/src/utils/burn_on_read_expiration_scheduler.ts

```diff
@@ -0,0 +1,316 @@
+     */
+    private checkAndExpirePosts(): void {
+        const now = Date.now();
+        const expiredPosts: string[] = [];
+
+        // Find all expired posts (including grace period in the check)
+        // This is where the grace period should be applied, not just in registerPost
+        for (const [postId, expireAt] of this.postExpirations.entries()) {
+            if (expireAt + this.gracePeriodMs <= now) {
+                expiredPosts.push(postId);
+            }
+        }
+
+        // Batch delete from tracking (avoids mutation during iteration)
+        for (const postId of expiredPosts) {
+            this.postExpirations.delete(postId);
+        }
+
+        // Dispatch batch expiration if dispatch is available
+        if (expiredPosts.length > 0 && this.dispatch) {
+            // Dispatch each expiration (in future, consider batched action to reduce Redux thrash)
+            for (const postId of expiredPosts) {
+                try {
+                    this.dispatch(handlePostExpired(postId));
+                } catch (error) {
+                    // Silently fail - errors are logged by Redux middleware
+                }
+            }
+        }
+
+        // Reschedule if we expired any posts
+        // - If we expired posts and more remain: schedule the next one
+        // - If we expired posts and none remain: recomputeSchedule() exits early
+        if (expiredPosts.length > 0) {
+            this.recomputeSchedule();
+        }
+    }
+
+    /**
+     * Handle visibility change - check expirations when tab becomes visible
+     */
+    private handleVisibilityChange = (): void => {
+        if (!document.hidden) {
+            // Tab became visible, check for expired posts
+            this.checkAndExpirePosts();
+
+            // Reschedule in case timing changed while tab was hidden
+            if (this.postExpirations.size > 0) {
+                this.recomputeSchedule();
+            }
+        }
+    };
+
+    /**
+     * Handle single post expiration (used by registerPost for immediate expiration)
+     */
+    private handleExpiration(postId: string): void {
+        // Remove from tracking
+        this.postExpirations.delete(postId);
+
+        // Dispatch expiration action
+        if (this.dispatch) {
+            try {
+                this.dispatch(handlePostExpired(postId));
+            } catch (error) {
+                // Silently fail - errors are logged by Redux middleware
+            }
+        }
+    }
+
+    /**
+     * Determine which timer is active (whichever expires first)
+     */
+    private getActiveExpiration(expireAt: number | null, maxExpireAt: number | null): number | null {
+        if (!expireAt && !maxExpireAt) {
+            return null;
+        }
+        if (!expireAt) {
+            return maxExpireAt;
+        }
+        if (!maxExpireAt) {
+            return expireAt;
+        }
+
+        // Whichever expires first
+        return Math.min(expireAt, maxExpireAt);
```

Comment:
Under no circumstance will the global expiry will be before the user expiry. If that happens, there are far more severe bugs. This check is unnecessary. No need of such unnecessary checks.

Reply:
since the expireAt depends on when the user reveals the message, then a valid scenario would be something like:

1. Post created at T=0 with max_ttl=15min → max_expire_at = T+15min
2. User reveals at T=10min with duration=10min → expire_at = T+20min  
3. Max TTL expires FIRST at T+15min (before reveal timer at T+20min)

then, the Math.min() will correctly pick whoever expires first, which is the intended behavior.